design/one-pager-change-logs.md

```diff
@@ -0,0 +1,310 @@
+correct this unexpected change without any user interaction with the control
+plane.  Currently, there is no mechanism for users of Crossplane to have
+visibility into these types of "unexpected" changes that Crossplane is making on
+their behalf.
+
+With Crossplane acting continuously and autonomously to update critical
+infrastructure, it is vital for users to have insight into the operations being
+performed, so they can build and maintain a strong sense of confidence and trust
+in their control planes.
+
+There are many approaches that can (and should) be taken to build trust along
+the various phases of the lifecycle of control planes and the infrastructure
+that they manage, but this design will focus specifically on the observability
+of changes that the control plane makes to external systems at runtime. This is
+not intended to facilitate review or approve the changes before they are made,
+but rather to provide an auditable record of all changes that have been
+performed.
+
+This means that Crossplane (and its Providers) should start to expose data about
+all changes that the control plane is making to external systems, with as much
+detail as possible so the platform team can understand what is changing, for
+what reason it is changing, and the outcome of the change. With this
+information, Crossplane users will be able to build deeper trust and confidence
+in the operation of their control planes, as well as have forensic data
+available to them in the event that an unexpected change occurs.
+
+## Proposal
+
+### Change Log Entry Format
+
+The change log will be a series of individual entries that capture state about
+the resources involved in every operation a provider performs on its
+external system. Later in this document we will explore HOW the logs are
+generated and stored, but for now we will focus on WHAT will be collected.
+
+As a general principal, we will be collecting raw observation data that has not
+been processed in any significant way. A system of the users choosing that
+collects the logs entries can perform opinionated calculations, analysis, or
+alerting that is specific to their operating environment.
+
+Each entry in the change log will contain the following data:
+
+| Field Description | Example |
+|-------------------| ------- |
+| Timestamp | ISO 8601 format, e.g. `2023-04-01T12:34:56Z` |
+| Provider name | `xpkg.upbound.io/upbound/provider-aws-ec2:v1.8.0` |
+| Managed Resource identifying data (GVK/name) | `{apiVersion: ec2.aws.upbound.io/v1beta2, kind: Instance, name: dev-instance-bt66d}` |
+| Operation Type | `create\|update\|delete` |
+| Before operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider` |
+| After operation desired/observed state of resource | `{full JSON dump of resource}` which includes desired `spec.forProvider` and observed `status.AtProvider`  |
```

Comment:
For providers returning the control back to the managed reconciler before the external operation completes (upjet does so in the default async reconciliation mode), we will in general record the same state as the `before` and the `after` state. We were planning to introduce modifications in the managed reconciler so that async reconciliation becomes a first class citizen in crossplane-runtime. Another option could be to give the control of logging to the provider itself, e.g., give a handle of the gRPC client to the provider, so that the upjet runtime can properly report the before-update & after-update states of the external resource.

We also discussed the utility of the `spec.forProvider` as the reconciler does not get a hold of the old spec in case it's been updated (this would especially be relevant if it's the spec update that will result in an external resource update and in turn, result in a change log). The only common exception is late-initialization but this is not relevant to the change logs discussion. So, the desired state will in general be identical in the before and after logs related to a single change.

Reply:
> We were planning to introduce modifications in the managed reconciler so that async reconciliation becomes a first class citizen in crossplane-runtime.

Is there somewhere I can learn more about these plans? I'd like to understand the async operation approach better. I understand that Terraform operations can take a long time to complete, but I've never personally been sold on the async operation approach.

Reply:
Thanks @ulucinar for writing down this observation about providers that perform async operations, it's a good point to understand. As I've been thinking through this design further, I've been increasingly thinking of the "after" state we capture as being a "best effort" approach.

Within a single `Reconcile()`, we can perform the operation and then call `Observe()` to get the latest state from the provider, with the understanding that the available state at this point in time may not yet reflect the operation we just performed as being concluded. If there are no further change logs for the given resource, we can perhaps assume that it converged with the desired state.

That may be acceptable IMO, at least for this initial alpha implementation 🤔

Reply:
> I understand that Terraform operations can take a long time to complete, but I've never personally been sold on the async operation approach.

Things might have changed with the latest changes in the upjet framework, especially with the no fork architecture, but AFAIR, a terraform apply call can take as long as the timeouts defined for a given terraform resource. For example, AWS [RDS instance](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/rds_cluster#timeouts) can take up to 120m and blocking the reconcile for long-running tasks is against controller [best practices](https://github.com/kubernetes-sigs/controller-runtime/issues/1763#issuecomment-1003703642). 

In provider-helm, we had a similar [need](https://github.com/crossplane-contrib/provider-helm/issues/63) and could live up to today without implementing it at the cost of not providing a great UX for some couple of rare cases where a helm install taking much longer (e.g. a pre-install hook, or wait=true). In case of terraform resources, this is more severe (i.e. longer).

Reply:
See https://github.com/crossplane/crossplane/pull/5822#discussion_r1701017542 for where we have landed on no longer collecting the "after" state by performing a 2nd `Observe()` call.